internal/uvm/network.go

```diff
@@ -1,31 +1,129 @@
 package uvm
 
 import (
-	"fmt"
+	"errors"
 	"path"
 
+	"github.com/sirupsen/logrus"
+
 	"github.com/Microsoft/hcsshim/hcn"
 	"github.com/Microsoft/hcsshim/internal/guestrequest"
 	"github.com/Microsoft/hcsshim/internal/guid"
 	"github.com/Microsoft/hcsshim/internal/hns"
+	"github.com/Microsoft/hcsshim/internal/logfields"
 	"github.com/Microsoft/hcsshim/internal/requesttype"
 	"github.com/Microsoft/hcsshim/internal/schema1"
 	"github.com/Microsoft/hcsshim/internal/schema2"
 	"github.com/Microsoft/hcsshim/osversion"
-	"github.com/sirupsen/logrus"
 )
 
-// AddNetNS adds network namespace inside the guest & adds endpoints to the guest on that namepace
-func (uvm *UtilityVM) AddNetNS(id string, endpoints []*hns.HNSEndpoint) (err error) {
+var (
+	// ErrNetNSAlreadyAttached is an error indicating the guest UVM already has
+	// an endpoint by this id.
+	ErrNetNSAlreadyAttached = errors.New("network namespace already added")
+	ErrNetNSNotFound        = errors.New("network namespace not found")
+)
+
+// AddNetNS adds network namespace inside the guest.
+//
+// If a namespace with `id` already exists returns `ErrNetNSAlreadyAttached`.
+func (uvm *UtilityVM) AddNetNS(id string) error {
+	logrus.WithFields(logrus.Fields{
+		logfields.UVMID: uvm.id,
+		"netns-id":      id,
+	}).Debug("uvm::AddNetNS")
+
+	uvm.m.Lock()
+	defer uvm.m.Unlock()
+	if _, ok := uvm.namespaces[id]; ok {
+		return ErrNetNSAlreadyAttached
+	}
+
+	if uvm.isNetworkNamespaceSupported() {
+		// Add a Guest Network namespace. On LCOW we add the adapters
+		// dynamically.
+		if uvm.operatingSystem == "windows" {
+			hcnNamespace, err := hcn.GetNamespaceByID(id)
+			if err != nil {
+				return err
+			}
+			guestNamespace := hcsschema.ModifySettingRequest{
+				GuestRequest: guestrequest.GuestRequest{
+					ResourceType: guestrequest.ResourceTypeNetworkNamespace,
+					RequestType:  requesttype.Add,
+					Settings:     hcnNamespace,
+				},
+			}
+			if err := uvm.Modify(&guestNamespace); err != nil {
+				return err
+			}
+		}
+	}
+
+	if uvm.namespaces == nil {
+		uvm.namespaces = make(map[string]*namespaceInfo)
+	}
+	uvm.namespaces[id] = &namespaceInfo{
+		nics: make(map[string]*nicInfo),
+	}
+	return nil
+}
+
+// AddEndpointsToNS adds all unique `endpoints` to the network namespace
+// matching `id`. On failure does not roll back any previously successfully
+// added endpoints.
+//
+// If no network namespace matches `id` returns `ErrNetNSNotFound`.
+func (uvm *UtilityVM) AddEndpointsToNS(id string, endpoints []*hns.HNSEndpoint) error {
+	logrus.WithFields(logrus.Fields{
+		logfields.UVMID: uvm.id,
+		"netns-id":      id,
+	}).Debug("uvm::AddEndpointsToNS")
+
 	uvm.m.Lock()
 	defer uvm.m.Unlock()
```

Comment:
Similar comment here about length the lock is held.

Reply:
Same